### ISSUE TITLE
sticky: unofficial: the extension csp header modification game [1462989 + 1635781 resolved 78 / ESR78.1]

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#664]
Anyone want a game of craps?

- **required reading**: #265 
- **bugzillas**: [1421725](https://bugzilla.mozilla.org/show_bug.cgi?id=1421725), ~~[1417249](https://bugzilla.mozilla.org/show_bug.cgi?id=1417249)~~ marked duplicate of [1477696](https://bugzilla.mozilla.org/show_bug.cgi?id=1477696), and [1462989](https://bugzilla.mozilla.org/show_bug.cgi?id=1462989)
- **short description**: when two or more extensions use CSP injection to modify headers on the same page, only one wins. It doesn't matter who: first loaded, first modified - don't care: the fact is only one extension will achieve what it is meant to, the other(s) will **fail**
- **FYI**: The wiki page for [extensions](https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.1-Extensions) has been updated to reflect just how dangerous this is (such as JS blocking not getting blocked), but is by no means definitive

Use this sticky to report where settings use CSP: especially in our recommended extensions - outside of those, I don't really care, but it would be interesting, for example, to know that other popular extensions such as ABP or even Chameleon cause issues

:star: CSP header injection is used extensively in uBlock Origin, for example (see gwarser's comment four posts down). The entries below are some features you could **consider** disabling (or achieving another way) to **reduce** CSP header injection issues, **depending on what extensions you use**.  It is up to you to determine what mix you want

template
```
- extension: 
- setting:
   * comment
- solution:
```

`TIP` from gwarser :kiss: [CRX viewer](https://addons.mozilla.org/en-US/firefox/addon/crxviewer/?src=search) and search for `!content-security-policy`. (`!` means scan all files.) If they touch this header, they are potentially unsafe.

**Comment by Thorin-Oakenpants** [comment]
extension: uBlock Origin
setting: Dashboard>Settings>Block remote fonts. This sets a font **rule** and font rules use CSP header modification [no word from gorhill if font filters are an issue, asked twice, I guess he's busy - @gwarser then)
alternative: use Request Control

**Comment by Thorin-Oakenpants** [comment]
- extension: Canvas Blocker
 - setting: `Misc > Block data URL pages`
 - solution: uncheck the option

**Comment by Thorin-Oakenpants** [comment]
- extension: HTTPS-Everywhere
- setting: uncheck toolbar icon dropdown> `Encrypt All Sites Eligible (EASE)`
   * by default this is unchecked
- solution: don't check it

**Comment by gwarser** [comment]
This will not be so easy in uBO. I don't think CSP can be safely disabled without modifying uBO code. Also, disabling CSP filtering will cripple uBO a lot.

Starting from here: https://github.com/gorhill/uBlock/blob/87feb47b51202cb8464eab91597b706965a224f3/src/js/traffic.js#L764

- Inline script filtering is powered by CSP (except for HTML filtering - `##^`, but this will not work for all scripts)
  - dynamic filtering - `example.com * inline-script block`
  - static filtering - `||example.com^$inline-script` (over 100 filters total)
  - "no scripting" feature - `no-scripting: * true`
- Inline fonts:
  - "Block remote fonts" feature - `no-remote-fonts: example.com true`. Related code: [1](https://github.com/gorhill/uBlock/blob/87feb47b51202cb8464eab91597b706965a224f3/src/js/traffic.js#L801) (called for `main_frame`), [2](https://github.com/gorhill/uBlock/blob/87feb47b51202cb8464eab91597b706965a224f3/src/js/pagestore.js#L563), [3](https://github.com/gorhill/uBlock/blob/87feb47b51202cb8464eab91597b706965a224f3/src/js/pagestore.js#L658) (Not related to `$font` network filtering option)
  - `inline-font` static filter option, but I cannot find even one in my dump of uBO compatible filters from filterlists.com
- Any filter list can use `$csp=` filters, and there are lot of them. ~400 in uBO "Filter lists", over 800 total.

**Comment by Thorin-Oakenpants** [comment]
OK, sorry. I thought this was a given: It's not about trying to remove CSP header modification - it's about **reducing** the possibility of two or more extensions trying to use it on the same page. We absolutely push uBO (and uM)

The above three examples are to reduce the problem:
- e.g font rules in uBO can be replaced by Request Control (that's what I did because as an example it blocked uM stopping workers in a test I did)
- e.g instead of HTTPS-E strict secure, you could enforce no mixed content via prefs to mitigate anything (not to mention uM to block all 3rd party by default etc to reduce requests)
- no idea about a solution for the CB one

^^ all these examples are so extensions still work as expected. As you said, uBO needs it

**Comment by tartpvule** [comment]
- extension: uMatrix
- settings:
  - Inline script blocking
  - Inline style blocking
  - "Forbid web workers"
- solution: do not use those features :( severe loss of functionality there :(

FYI, if it is ever used by security extensions, [Feature-Policy](https://bugzilla.mozilla.org/show_bug.cgi?id=1390801) will have much of the same problems as today's CSP.

/rant start/ The writing is on the wall for Chromium, Google fully intends to cripple user's choices, just look at Manifest v3 and their extremely limited declarativeWebRequest junk. Their "ad blocker" is just a "good enough" junk for unsuspecting users so they don't go seek out more effective, real privacy-protecting blockers. Hope Mozilla will not try to mimic Chromium's insanity, but that seems to be the trend nowadays. /rant end/

**Comment by Thorin-Oakenpants** [comment]
@tartpvule I don't know what to do with that. I'm not trying to tell everyone to disable all CSP header injection, but instead selectively point out where there can be conflicts, and try to provide a workaround. And each item can be a different "threat", e.g depending on how often the clash happens: e.g an inline script block curated for a single domain may never cause a conflict.

I'm not an expert on uM or uBO - are you saying that when uM uses any of those three things, that it uses CSP header injection? I modified your post, btw

**Comment by atomGit** [comment]
@Thorin-Oakenpants said (on my website):

> *$font,third-party , uBO section: don’t: it can cause CSP issues

uBO and/or filter lists apparently do a lot of header mods already, so i don't understand why setting that filter would raise issues - ELI5 :)

**Comment by Thorin-Oakenpants** [comment]
Sorry, is that a **filter**? I have no idea about filters, I asked gorhill twice and gwarser once - never got an answer that I know of. The font **rule** uses CSP.

**Comment by atomGit** [comment]
> Sorry, is that a filter?

yessir, if you add that to the 'My Filters' list it allows 1st part fonts globally (aka rejects 3rd party fonts globally) - i'll also drop this here should anyone find this while looking for info on that filter...

to allow 3rd party font for a domain:
`*$font,third-party,domain=~example.com`
to allow for multiple domains:
`*$font,third-party,domain=~example.com|~example2.com`

**Comment by crssi** [comment]
> if you add that to the 'My Filters' list it allows 1st part fonts globally

Not really. 1st party is newer mentioned in that line.

**Comment by Thorin-Oakenpants** [comment]
@atomGit , @crssi can you just delete your comments. I **hate** these stickies getting filled up with conversation. I really just want them to be a simple list of things for people to be aware of.  - thanks

I shall of course, delete this comment later on

**Comment by atomGit** [comment]
it seems to work for me (both allow 1st and block 3rd) and that code came from [Gorhill](https://github.com/gorhill/uBlock/issues/363#issuecomment-191796040) and IDKwhattoputhere

**Comment by gwarser** [comment]
>Sorry, is that a filter? I have no idea about filters, I asked gorhill twice and gwarser once - never got an answer that I know of. The font rule uses CSP.

Ekhem... https://github.com/ghacksuserjs/ghacks-user.js/issues/664#issuecomment-472596147

- Inline fonts:
  - "Block remote fonts" feature - `no-remote-fonts: example.com true`. Related code: [1](https://github.com/gorhill/uBlock/blob/87feb47b51202cb8464eab91597b706965a224f3/src/js/traffic.js#L801) (called for `main_frame`), [2](https://github.com/gorhill/uBlock/blob/87feb47b51202cb8464eab91597b706965a224f3/src/js/pagestore.js#L563), [3](https://github.com/gorhill/uBlock/blob/87feb47b51202cb8464eab91597b706965a224f3/src/js/pagestore.js#L658) (Not related to `$font` network filtering option)
  - `inline-font` static filter option, but I cannot find even one in my dump of uBO compatible filters from filterlists.com


You can remove this comment.

**Comment by curiosity-seeker** [comment]
>     * solution: use Request Control (or maybe font filters in uBO? there's a uBO wiki page about that - if I can find it and it's relevant)

I think you mean [this](https://github.com/gorhill/uBlock/wiki/Per-site-switches#no-remote-fonts) one. It doesn't answer the question if font filters cause this problem as well or not, though.

**Comment by atomGit** [comment]
is it enough to search the ext. code for `!content-security-policy` with CRX to verify whether an ext. is using CSP?

i ask because in da wiki, for example, ClearURLs is listed as using CSP, but the CRX search doesn't find `content-security-policy`

**Comment by Thorin-Oakenpants** [comment]
In whose wiki. In our one it doesn't say that about ClearURLs at all

**Comment by atomGit** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.1-Extensions

**Comment by Thorin-Oakenpants** [comment]
Where does it say ClearURLs is using CSP?

![where-csp](https://user-images.githubusercontent.com/16656956/57339586-bec5bc00-7121-11e9-9e36-95918c3bf189.png)

**Comment by atomGit** [comment]
sorry Pants - my mistake - because of the indentation of the bullet list, i thought all those extensions were sub headings of the CSP paragraph - i missed the association of the red ! with the extensions that use CSP

**Comment by Thorin-Oakenpants** [comment]
@curiosity-seeker Yeah, that the link that explains the alternative, but doesn't answer my question. But gwarser did in post 5 (see the inline fonts bit), but I have no idea how to interpret that - I'm not an expert

**Comment by gwarser** [comment]
CSP is used only for blocking inline fonts (base64 encoded). Inline fonts are blocked only when "No remote fonts" popup switch is used, or if specified in static filter option `$inline-font`. `$font` option block by _classic_ web request blocking.

**Comment by curiosity-seeker** [comment]
@gwarser : Just to make this perfectly clear: `$font` does **not** use CSP, and `$inline-font` does **not** use CSP, either. Correct?

**Comment by atomGit** [comment]
not correct - gwarser is saying that `$inline-font` when used in a static filter, such as the filter subscriptions, DOES use CSP - i just checked some of the bigger filter lists (easy, adguard, fanboy and the uBlock filters) and none use `$inline-font`

**Comment by Thorin-Oakenpants** [comment]
I'm just confused. What do I edit the info in post 2 to?

**Comment by Aeriem** [comment]
Hello there,

I'm sorry for the months of silence, life found its way to catch up to me and stomp me hard.
HTTPS-Everwhere's team is now aware of the CSP issue ([HTTPS-E conflicts with NoScript in Tor Browser at Safest security setting when the EASE mode is enabled #17735](https://github.com/EFForg/https-everywhere/issues/17735)).
I'm hoping that since this affects TBB, Firefox's developers will see how critical this issue actually is.
Much luv' and congrats on the spring cleaning! It was so easy to get back in the privacy game thanks to your efforts on the documentation and pref cleaning, I'm baffled. Overrides also are much fewer than they used to be. :D

**Comment by pipboy96** [comment]
This is why there should be an official way to apply CSP and FP to web pages that's standardised, instead of current hackish way.

**Comment by Aeriem** [comment]
I completely agree.
Hopefully this will be resolved in the near future.
Can anything be done to help?

**Comment by Thorin-Oakenpants** [comment]
> Can anything be done to help?

I'm working on it from **within**

**Comment by Aeriem** [comment]
Imma scared, did you infiltrate Mozilla's headquarters? 😱
Blink twice if you need us to exfiltrate you quickly!

I'm curious though, are you working with the devs from the Bugzilla tickets you linked? Because that'd be super cool.

**Comment by Thorin-Oakenpants** [comment]
FukYeah™ - Jame Bond shit

I *almost* typed something I shouldn't (personally) - not that I care (I've  never applied OpSec here really or tied to hide who I am: but I know how). I have spent years working my way in (its like a game/passion to me (learning and my passion (privacy, anonymity) , not "infiltrating", that's just a side effect: nothing like a spy or anything) and to try to have a voice (nothing nutty of course: I can see all sides (its one of my strengths even though I'm a selfish bastard in real life). I have *curated* a large'ish list of **top** Mozilla and some **top** Tor Project  people (I always feel out of my depth with Tor people, and never tried - but fuck, they're awesome: I should have tried years ago) - to the point where I am actually ~~being asked for my opinion~~ making a difference (thru perseverance: maybe they just put up with me: although I brought thing to their attention they (probably maybe) never thought of: yes, I am directly responsible for starting some fo the changes you already enjoy)) - you have been enjoying my input for a while now ( a little input) . Now that (in the last 6 months, due to personal circumstances, I no longer care about anything except making the world "better" before I go) I have actually focused on ONE thing ... its all ON baby, full tilt (IMO). I stepped my game up: it's my mission: that mission is not CSP: its something bigger (but CSP is on there)

If you can parse that, well done

PS: disclaimer: right now I am almost naked and drunk AF (why wouldn't you be, seriously: best shit ever)

PPS: personal circumstances: no I am not terminal, so STFU, I lost my best ever buddy and got (and still am) severely depressed; but I have a mission and that keeps me going (for now)

PPPS: sorry if I come across as a dick

**Comment by Thorin-Oakenpants** [comment]
^^ did I say I was drunk and naked

**Comment by pipboy96** [comment]
@Thorin-Oakenpants Um, everything's fine? Should we be scared?

**Comment by Thorin-Oakenpants** [comment]
WTF would you ask me that when I'm naked? Everything is cool!

**Comment by Aeriem** [comment]
Must admit I had to reread some sentences to fully get you haha!
But I'm no better with mine, so who am I to judge? 🤷

I feel like the conversation is getting a bit personal and out of the scope of this git issue, but I'd love to carry it on somewhere else (if you don't mind).
Hit me up if you want to.
Still, I'm sorry for your loss and your work is appreciated. I learned so much from this repo.

@pipboy96 Thank you again for taking the time to review this issue, and sorry for the flood. Also, Pants quirky personality is one of the reasons this repo is quite fun to read.

**Comment by crssi** [comment]
Yeah, maybe @Thorin-Oakenpants is not aware of, but a most of times the writing is real fun to read.
Sorry for the loss.

**Comment by pipboy96** [comment]
@Thorin-Oakenpants I'm sorry for your loss. May your friend rest in peace.

**Comment by atomGit** [comment]
sorry about your loss Pants - wish i had some magic words that would help, but i never know what to say in circumstances like this

> Can anything be done to help?

vote for [1421725](https://bugzilla.mozilla.org/show_bug.cgi?id=1421725)

**Comment by atomGit** [comment]
what happens if `content_security_policy` is contained in the manifest of an add-on, but `content-security-policy` isn't contained anywhere?

`"content_security_policy": "default-src 'self'; connect-src ws:; style-src 'self' 'unsafe-inline'"`

**Comment by gwarser** [comment]
This applies only to extension https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy

Your example is bit unusual with (not encrypted) websocket and inline stylesheets.

**Comment by atomGit** [comment]
ok, i understand now - [Echo for Kodi](https://addons.mozilla.org/en-US/firefox/addon/echo-for-kodi/) is the ext. i'm referring to btw

**Comment by Giltyhub** [comment]
Relevant: https://bugzilla.mozilla.org/show_bug.cgi?id=1462989

**Comment by Thorin-Oakenpants** [comment]
Its already listed in OP

**Comment by WellOrientedLlama** [comment]
Just for the record, it seems that the last add-on to be loaded by Firefox wins the race and gets its HTTP header changes through, crushing other extensions' tweaks.

So my Firefox starts with the page `about:addons`, and on each startup and each add-on update, I disable uMatrix and reenable it again, so it is always the last add-on to be loaded by the Fox. Then it wins against uBlock Origin's relevant features which is what I wanted.

Additionally I...don't think it's limited to CSP ? Shouldn't any HTTP header be affected ? uBO's `blob:` or `data:` filters are also affected, if I recall correctly (been a while), though I don't know how uBO translates them internally.

**Comment by Atavic** [comment]
> Additionally I...don't think it's limited to CSP ? Shouldn't any HTTP header be affected ?

Any addon race that's conflicting on the same header's modification acts the same.

**Comment by earthlng** [comment]
> Additionally I...don't think it's limited to CSP

no it's not limited to CSP but CSP is the most important one if you care about reliably blocking inline Javascript.

> Shouldn't any HTTP header be affected

yes any HTTP header is affected if 2 or more addons try to modify the same header

>  uBO's `blob:` or `data:` filters are also affected

yes but AFAIK uBO blocks or limits those by using a CSP header

**Comment by kah0922** [comment]
Well, it seems like shit is actually being worked on now in [bug 1262989](https://bugzilla.mozilla.org/show_bug.cgi?id=1462989).

**Comment by travankor** [comment]
Alternatively, recompile Firefox with the patch from this repo https://github.com/tartpvule/my-firefox-patches (see #265, except it's closed)

I'm disappointed that neither Tor nor mozilla care about this issue -- maybe it's an intentional loophole to make it easier to identify Tor browser users?

**Comment by andretiagogr** [comment]
> I'm disappointed that neither Tor nor mozilla care about this issue -- maybe it's an intentional loophole to make it easier to identify Tor browser users?

Tor doesn't recommend installing extensions and NoScript and HTTPS Everywhere won't conflict with each other because EASE is disabled by default. That's probably why Tor doesn't care

**Comment by KOLANICH** [comment]
>EASE is disabled by default

That's shit.

**Comment by rusty-snake** [comment]
> > EASE is disabled by default
> 
> That's shit.

Default security level: Standart
Reason: compatibility

**Comment by atomGit** [comment]
just a heads up on the CSP issue :)

[Status: NEW → ASSIGNED](https://bugzilla.mozilla.org/show_bug.cgi?id=1462989#a55402279_600971)

**Comment by Thorin-Oakenpants** [comment]
I know. I've kept a close eye on it since day 1. I've actually met Shane (mixedpuppy). Anyway, I created a new [label](https://github.com/ghacksuserjs/ghacks-user.js/labels) to apply to #265 and this ticket when it gets done

![jesus](https://user-images.githubusercontent.com/16656956/74988415-aff09180-5435-11ea-9a25-0275169940a0.png)

It's also going to get used on #448 because a solution for that is actually already available, it just needs some perf work before it lands (not going to say anymore since the tickets are access denied)

**Comment by atomGit** [comment]
[support merging content-security-policy headers provided by multiple extensions](https://bugzilla.mozilla.org/show_bug.cgi?id=1462989)
Status: RESOLVED FIXED
Milestone: mozilla77

anybody know what "fixed" means? i looked at the latest activity and i have no idea how this was fixed or what the fix translates to in terms of how extensions are affected

**Comment by gwarser** [comment]
Previously, when original response had CSP header and extensions tried to set another, extension api had knowledge of original (from server) header and was able to set  `headersAlreadySet` flag which instruct Firefox to merge headers.

When original response did not had CSP header, extensions running in different _threads_ "at once" did not have chance to read `headersAlreadySet` set by other extension, so all extensions thought that they have only header and they do not instruct Firefox to merge headers - only one "win".

Now `headersAlreadySet` for CSP is [pre-set / hardcoded](https://hg.mozilla.org/mozilla-central/rev/cea45f637b49), so extensions headers will always be merged.

**Comment by atomGit** [comment]
thank you for that explanation, but it's a bit over my head technically

currently i have 3 extensions that make use of CSP, one being uMatrix

in order to ensure that uM takes precedence, every time i start FF i must disable, then enable uM so that it becomes the last (or first?) in the chain, else in-line JS may not be blocked

so my question is, would one still need to do this in FF 77?

given your explanation, i'm guessing the answer is 'no', but i'm not sure

**Comment by Thorin-Oakenpants** [comment]
I'll await anointing any `jesus was listening` labels until it's all been tested. My understanding is that there are still potential issues e.g. if two headers clash- but IANAE

**Comment by uBlock-user** [comment]
That bug specifically fixes CSP header clashes with a force merge, nothing else. All other headers may still clash.

**Comment by travankor** [comment]
Is there a tl;dr of what we need for a comprehensive fix (ie: what's left to do)?

**Comment by Thorin-Oakenpants** [comment]
> Is there a tl;dr of what we need for a comprehensive fix (ie: what's left to do)?

- test: assigning gorhill, gwarser, earthlng, and uBlock-user : feel free to participate
- investigate (for peace of mind), update documents: e.g.
   - I for one would like to know how to compares to header merging pre-WebExt
   - what pitfalls still exist (maybe they always existed)
- wait for it to land in stable
- assign `jesus was listening` label
- get naked and drink beersies (optional)

**Comment by curiosityseeker** [comment]
> That bug specifically fixes CSP header clashes with a force merge, nothing else. All other headers may still clash.

Okay - but since the problem we're discussing here is about CSP headers (and not about other headers) it should be fixed now, shouldn't it?

**Comment by uBlock-user** [comment]
Fixed in Nightly yes.

**Comment by girst** [comment]
Hi, I'm the guy who wrote the patch. Oof, what a journey! If you want a 
pracical easy-to-repeat test, try this:

1. install uBlock Origin (optional, for 2nd and 3rd tests)
2. install uMatrix
3. install HTTPS Everywhere (order matters!)
    alternatively, disable and reenable the extensions in the above order
    to make sure their CSPs have the expected precedence
4. navigate to http://girst.2ix.at/tmp/webfont.html
    observe:
    - it will not redirect to https://...
    - "Javascript is enabled"
    - "Webfonts are [][][][][][][][]" (enabled)
5. block javascript using umatrix and refresh page
    observe:
    - "Javascript is disabled"
6. enable "encrypt all sites" in https everywhere and refresh page
    observe in nightly:
    - you are redirected to https://...
    - "Javascript is disabled"
    observe in release:
    - you are redirected to https://...
    - "Javascript is enabled" (!!!)

optional 2nd test:
7. enable font-blocking in uBlockO
8. using ctrl-f5, clear the cache and refresh the page
    observe in nightly:
    - "Webfonts are disabled"
    observe in release:
    - "Webfonts are [][][][][][][][]" (enabled) (!!!)

optional 3rd test:
9.  disable and reenable uMatrix
10. disable and reenable uBlockO
11. reload page
     observe in nightly:
     - everything still works
     observe in release:
     - "Javascript is enabled" (!!!)
     - "Webfonts are disabled"
     - unexpectedly, upgrade-insecure-requests still works (maybe some 
       browser caching going on?)

**Comment by uBlock-user** [comment]
@girst Thanks, without you, this bug would have never been fixed. Does this issue still exist for lets say Feature-Policy or any other header ?

**Comment by girst** [comment]
On Thu, Apr 16, 2020 at 07:09:11AM -0700, Saitama wrote:
>Does this issue still exist for lets say Feature-Policy or any other 
>header ?

yes. i submitted 2 variations of the patch, one that would apply to all 
headers, and one that only applies to c-s-p. mozilla/mixedpuppy went 
with the second one.

however, adding Feature-Policy to the exception "must-merge" list would 
be pretty easy: just add it to the headersAlreadySet set (and add a 
test, of course).

**Comment by travankor** [comment]
[According to security expert Daniel Micay](https://old.reddit.com/r/firefox/comments/gokcis/firefox_is_insecure_refuted/frlmdwu/), firefox "fails open" when it comes to web extensions?

@uBlock-user @girst 

Is this relevant to the discussion? If UBO dies from OOM, will firefox allow arbitrary connections?

**Comment by girst** [comment]
On Sun, May 24, 2020 at 03:09:23AM -0700, travankor wrote:
>[According to security expert Daniel 
>Micay](https://old.reddit.com/r/firefox/comments/gokcis/firefox_is_insecure_refuted/frlmdwu/), 
>firefox "fails open" when it comes to web extensions?

did some tests: yes that seems to be the case D:

i do understand why firefox does it this way (not every extension is 
designed to block requests, and the browser would be completely unusable 
if it were to block all network requests when an extension dies)

>Is this relevant to the discussion?

well, not directly to the topic of this thread

>If UBO dies from OOM, will firefox allow arbitrary connections?

yes. there are some steps to reproduce, with today's nightly:

1. install umatrix and ublocko
2. navigate to https://gir.st/tmp/webfont.html
3. in ublock, block remote fonts. in umatrix block javascript
4. reload with ctrl-f5 and notice requests are blocked
5. in a seperate tab, go to about:memory and hit "Measure"
    on the right, you'll see a list of all of fx's processes. note the
    one called "WebExtensions" and its process id.
6. open a terminal and issue `kill $PID_OF_WEBEXTENSION_PROCESS`
7. click on Measure again. the WebExtensions process should be gone. (if
    it isn't, kill -9 it, maybe)
8. reload with ctrl-f5 and notice requests are now going through.
    the browser console / terminal i started fx from will also fill with
    errors.

as far as i can tell, the webextension process is _not_ restarted 
automatically.  after a few minutes, i restarted the browser to get it 
back.

in my opinion, that last part, is the actual bug: if the webextension 
process dies, it should restart automatically, or at the very least warn 
the user. right now, the extension icons still show in the chrome, and 
only when clicking on them you'll notice that the extension doesn't 
respond.

looking at BMO, this appears to already be tracked: 
https://bugzilla.mozilla.org/show_bug.cgi?id=1355239 -- and it even has 
an assignee since a few days :D

**Comment by girst** [comment]
digging around in the bugs mentioned in the one from the last email: 
this comment[1] mentions `extensions.webextensions.remote`. setting it 
to false runs webextensions from "the parent/main process"[2].
there also appears to be a 
`extensions.webextensions.background-delayed-startup` pref that might be 
interesting[3] to some of you (i haven't found much documentation, other 
than a note that it is not going to stick around forever).

[1]: https://bugzilla.mozilla.org/show_bug.cgi?id=1627139#c8
[2]: https://searchfox.org/mozilla-central/rev/501eb4718d73870892d28f31a99b46f4783efaa0/modules/libpref/init/StaticPrefList.yaml#3343
[3]: https://searchfox.org/mozilla-central/rev/501eb4718d73870892d28f31a99b46f4783efaa0/browser/app/profile/firefox.js#70

**Comment by girst** [comment]
apologies for the spam (last one, i promise).

some caveats regarding webextensins.remote:

> As Andrew noted, extensions.webextensions.remote is not really 
> supported. Its removal was planned in bug 1613141, but I'll hold off 
> from doing so until we have a decent recovery strategy for extension 
> process crashes (especially since we cannot remove all related code 
> until we support out-of-process extensions on mobile).

https://bugzilla.mozilla.org/show_bug.cgi?id=1627139#c10

**Comment by Thorin-Oakenpants** [comment]
indeed .... `https://github.com/uBlockOrigin/uMatrix-issues/issues/231#issuecomment-596062505`

**Comment by geeknik** [comment]
My takeaway here is to build defense in depth and not just rely on the browser to protect you. Since we run pfsense on our router locally, we use pfblocker-ng to do the heavy lifting and UBO to do the surgical cleaning at the user level. Running [Pi-Hole](https://pi-hole.net/) is another option. Good luck out there!

**Comment by curiosityseeker** [comment]
> yes. there are some steps to reproduce, with today's nightly: 1. install umatrix and ublocko 2. navigate to https://gir.st/tmp/webfont.html 3. in ublock, block remote fonts. in umatrix block javascript 4. reload with ctrl-f5 and notice requests are blocked 5. in a seperate tab, go to about:memory and hit "Measure" on the right, you'll see a list of all of fx's processes. note the one called "WebExtensions" and its process id. 6. open a terminal and issue `kill $PID_OF_WEBEXTENSION_PROCESS` 7. click on Measure again. the WebExtensions process should be gone. (if it isn't, kill -9 it, maybe) 8. reload with ctrl-f5 and notice requests are now going through. the browser console / terminal i started fx from will also fill with errors. as far as i can tell, the webextension process is _not_ restarted automatically. 

Thanks, I could reproduce. This should not happen. However, the question is: how relevant is this in practice? I've never seen my extensions dying. But obviously it does happen according to the uMatrix issue mentioned by Pants.

> looking at BMO, this appears to already be tracked: https://bugzilla.mozilla.org/show_bug.cgi?id=1355239 -- and it even has an assignee since a few days :D

Good. If Rob Wu takes care of it a solution should not be far away.

**Comment by girst** [comment]
On Sun, May 24, 2020 at 07:43:09AM -0700, curiosityseeker wrote:
>Thanks, I could reproduce. This should not happen. However, the 
>question is: how relevant is this in practice? I've never seen my 
>extensions dying. But obviously it does happen according to the uMatrix 
>issue mentioned by Pants.

I guess you have to be both a tab hoarder and only limited amount of 
memory available. I can't say I personally seen it happen on any of my 
machines.

If you're worried you might be affected, I cobbled together a small 
shell script that can warn you whenever the process dies.

     tail --pid=`grep -H WebExtensions /proc/*/comm | awk -F/ '{print $3}'` -f /dev/null &&
       zenity --info --text 'WebExtension process died!'

It's a bit dense, but here's how it works:
- we search all available processes' comm file for the name of the 
   process, "WebExtensions", and extract the PID (grep -H includes the
   path in the output, which we split on '/' and extract the PID w/ awk)
- tail --pid (another gnuism) waits for a process with said pid to exit
- once tail exits, pop up a info box using zenity. this could of course 
   be replaced by anything you'd want to do when this happens (e.g.  kill
   firefox).

**Comment by Thorin-Oakenpants** [comment]
This OOM / extention-hard-fail is OT, but I thought I would share, because sharing is caring - anyone want a hug? I haven't broken my solitude or left my shack in the remote woods for eight weeks, and I promise to shower with lots of soap first. Any takers? I'll be gentle.

@geeknik mentioned, in a now deleted post (why do you keep deleting posts right after you make them), defense in depth, about using something like PiHole on the router. Sure. You could also use the OS level.

Now I was thinking about Tor Browser users, and how TB doesn't ship with any adblocking (except tails). And this can be detected, indeed, even specific rules. All TB users on Tails should look the same in this regard, assuming that the default lists are static, lists used are not changed, custom filters/rules aren't added, etc.

IANAE on this: Does PiHole or similar affect your Tor Browser fingerprint. I mean, you're not the one making the final request for the content, so does it still come through, or is it blocked?

Besides that: when/if TB ships with some sort of adblocker (needs a lot of work, investigating: but gk is totally against blocking ads since it enables content: i.e it would also hurt the little guys): imagine the savings on the Tor network :)

**Comment by Thorin-Oakenpants** [comment]
FYI: https://trac.torproject.org/projects/tor/ticket/34305 .. I wonder if that's the trick gorhill mentioned about how NS always ensured it got the last CSP modfication

**Comment by curiosityseeker** [comment]
> I guess you have to be both a tab hoarder and only limited amount of memory available. 

Indeed, I don't belong to that group.

> tail --pid=`grep -H WebExtensions /proc/*/comm | awk -F/ '{print $3}'` -f /dev/null && zenity --info --text 'WebExtension process died!' 

Cool ! I don't think that I need it - but I've learned something new from it. Thanks!

**Comment by curiosityseeker** [comment]
> Does PiHole or similar affect your Tor Browser fingerprint. I mean, you're not the one making the final request for the content, so does it still come through, or is it blocked?

It is blocked. FWIW, I'm not using PiHole but dnscrypt-proxy with a huge blocklist. I'ver just tested it on https://browserleaks.com/proxy with uBO disabled. It didn't detect any content filters or filter subscriptions (which it does if uBO is enabled). Hence, the TB fingerprint should not be affected. Although that's certainly not a guarantee: If something is blocked by that blocklist which is also blocked by one of the well-known filterlists, it might be misinterpreted by the code running on such sites.

However, needless to say that solutions like PiHole, dnscrypt-proxy or AdGuardHome block adservers and tracking/malware domains on the DNS level very efficiently. But they are not able to block something like, e.g.,

`https://a.nytimes.com/svc/nyt/data-layer?sourceApp=nyt-vi&referrer=&assetUrl=https%3A%2F%2Fwww.nytimes.com%2F&`

which is blocked in uBO by the following rule: `||nytimes.com^*/data-layer?` So both blocking approaches are not really comparable.

**Comment by travankor** [comment]
I'm on FF 77.0.1 and the parent issue is still there. I'm using an old profile and disabled and re-enabled all extensions. I will try this on a new profile later on.

**Comment by Thorin-Oakenpants** [comment]
I wouldn't have thought the profile would make any difference. I was about to clean up the ext wiki page and assign the `jesus was listening` label and close this - but I'll hold off

**Comment by girst** [comment]
On Fri, Jun 05, 2020 at 05:54:42AM -0700, travankor wrote:
>I'm on FF 77.0.1 and the parent issue is still there. I'm using an old 

How did you determine if it's still broken? If you can give me your STR 
and list of extensions I'll have a look.

**Comment by travankor** [comment]
I tried the test you added [here](https://github.com/ghacksuserjs/ghacks-user.js/issues/664#issuecomment-614674122) on a clean profile (no user.js) and it fails everything.

In the console, I have `TypeError: setting getter-only property "suspended"`.

I think this might be unrelated to the bug in this thread.

**Comment by girst** [comment]
On Fri, Jun 05, 2020 at 08:12:58AM -0700, travankor wrote:
>I tried the test you added [here](...) on a clean profile (no user.js) 
>and it fails everything.

I just re-did the test. After you deactivate and reactivate the addons, 
you'll need to re-disable webfonts/javascript. It looks like, if you 
don't save those changes, they get reset when you disable the addon 
(which makes sense, in hindsight).

>In the console, I have `TypeError: setting getter-only property "suspended"`.
>I think this might be unrelated to the bug in this thread.

can't reproduce this.

**Comment by travankor** [comment]
This took an embarrasingly long time to figure out (although I realized as soon as I took a break and came back to this), but I had an obsolete `autoconfig.js` file in my firefox root directory.

I was under the impression that my package manager would have cleaned up these obsolete files a while ago, but apparently this is not the case... moral of the story: trust but verify.

**Comment by travankor** [comment]
>you'll need to re-disable webfonts/javascript. It looks like, if you
don't save those changes, they get reset when you disable the addon

Note that user.js has a option that disables webfonts, too (or, at least, your test fails with this option set). However, it seems like UBo blocks content than the native FF option.

**Comment by gwarser** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1635781 "Cannot replace frame-ancestors directive of the Content-Security-Policy HTTP Header via the webRequest API" was fixed.

**Comment by Thorin-Oakenpants** [comment]
I'm a little lost .. can someone ELI5 - this is iframes only right? Also, not in ESR78 which sucks: wiki will probably need updating once I understand better

**Comment by girst** [comment]
On Thu, Jun 25, 2020 at 02:18:39AM -0700, travankor wrote:
>cc @girst

no need to ping me, i'm still cc'd to this bug ;-)

it looks like this breaks the fix, according to a quick test i did. i 
already commented on the ticket on bmo with a shortened version of the 
str i provided to this thread earlier.

https://bugzilla.mozilla.org/show_bug.cgi?id=1635781#c18

**Comment by girst** [comment]
well, embarrassing: repeating the test made it work again. :|

(and there /is/ a regression test for it...which i wrote. and it 
evidently passed that one on check-in)

**Comment by Thorin-Oakenpants** [comment]
So this is "fully" resolved now in 78+ and ESR78.1 (when it comes out)

**Comment by jawz101** [comment]
Does this mean [this wiki page](https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.1-Extensions) is no longer relevant?  I'd rather not have to think about all of the factors it suggests to toggle in each extension or whether NoScript is not a recommended addon if the issue is now a non-issue.

**Comment by Thorin-Oakenpants** [comment]
if you're on stable, then the CSP stuff doesn't apply to you. Once ESR68 reached end-of-life, I will remove all the CSP warning
